### PR TITLE
Fixes Mechs two-shotting carp rifts

### PR DESCRIPTION
**Pull request by carlarctg**
## About The Pull Request

Fixes Mechs two-shotting carp rifts, by adding a get_mult proc that returns 0.75 on carp rifts (multiplier is 3 by default, so 4 times weaker).

## Why It's Good For The Game

Two-shotting bad. Additionally it's pretty easy to just facetank the dragon and slam the rift even if it's several times so I lowered the multiplier to 0.75, giving the dragon and le fishe more fighting chance.

## Changelog

:cl:
fix: Fixes Mechs two-shotting carp rifts, they're now considerably slower in breaking them with their fists.
/:cl:

### CODE REVIEW

**Comment by MrDoomBringer** [comment]
Please make your changelog make sense to players, someone reading that would have no idea what that means. 
e.g. "Mechs now take 8 punches to destroy carp rifts, down from 2"

**Comment by carlarctg** [comment]
a

File: code/modules/vehicles/mecha/mech_melee_attack.dm [hunk 1]
```diff
@@ -42,7 +42,13 @@
 			return 0
 	mecha_attacker.visible_message(span_danger("[mecha_attacker] hits [src]!"), span_danger("You hit [src]!"), null, COMBAT_MESSAGE_RANGE)
 	..()
-	return take_damage(mecha_attacker.force * 3, mecha_attacker.damtype, "melee", FALSE, get_dir(src, mecha_attacker)) // multiplied by 3 so we can hit objs hard but not be overpowered against mobs.
+	return take_damage(mecha_attacker.force * get_obj_mecha_damage_multiplier(), mecha_attacker.damtype, "melee", FALSE, get_dir(src, mecha_attacker)) // multiplied by 3 so we can hit objs hard but not be overpowered against mobs.
+
+/obj/proc/get_obj_mecha_damage_multiplier()
+	return 3
```
Comment (MrMelbert) [inline code/modules/vehicles/mecha/mech_melee_attack.dm:48]: Auto-doc the proc

File: code/modules/vehicles/mecha/mech_melee_attack.dm [hunk 2]
```diff
@@ -42,7 +42,13 @@
...
+
+/obj/structure/carp_rift/get_obj_mecha_damage_multiplier()
+	return 0.75
```
Comment (MrMelbert) [inline code/modules/vehicles/mecha/mech_melee_attack.dm:51]: Let's move this to the appropriate file, may as well.

File: code/modules/vehicles/mecha/mech_melee_attack.dm [hunk 3]
```diff
@@ -42,7 +42,13 @@
 			return 0
 	mecha_attacker.visible_message(span_danger("[mecha_attacker] hits [src]!"), span_danger("You hit [src]!"), null, COMBAT_MESSAGE_RANGE)
 	..()
-	return take_damage(mecha_attacker.force * 3, mecha_attacker.damtype, "melee", FALSE, get_dir(src, mecha_attacker)) // multiplied by 3 so we can hit objs hard but not be overpowered against mobs.
+	return take_damage(mecha_attacker.force * get_obj_mecha_damage_multiplier(), mecha_attacker.damtype, "melee", FALSE, get_dir(src, mecha_attacker)) // multiplied by 3 so we can hit objs hard but not be overpowered against mobs.
```
Comment (MrMelbert) [inline code/modules/vehicles/mecha/mech_melee_attack.dm:45]: Move this comment to the new proc. 
`return 3 // multiplied by 3 by default so we can hit objs hard but not be overpowered against mobs.`

File: code/modules/vehicles/mecha/mech_melee_attack.dm (hunk 1 above)
Comment (MrMelbert) [inline code/modules/vehicles/mecha/mech_melee_attack.dm:48]: Let's be nice to future consumers and pass the mecha and attacker to the proc. 
```
/obj/proc/get_obj_mecha_damage_multiplier(obj/vehicle/sealed/mecha/mecha_attacker, mob/living/user)
```

**Review by MrMelbert — CHANGES_REQUESTED** [review]

**Comment by Jackal-boop** [comment]
this is a balance change not a fix?

**Comment by moocowswag** [comment]
Hulk twoshotting portals being fixed was also considered and merged as a balance change but I dont agree with it given that the portals were not meant to be easy to destroy its just that they happen to be susceptible to anything that does high object damage.


Which is the real meat of the issue, alot of objects and structures have powercreeped stats (tons of armor/dmgresist) to survive a few hits from alot of entities that do high object damage, its why I`m having trouble balancing my forcefields because I cant make them decent against alot of the mobs that have bloated object damage without making them nigh unbreakable to regular crew.



Speaking of mobs, I`m pretty sure specific ones made sentient via xenobio (especially when giving the strength / size boosting organ thingy) will be able to quickly murder portals too.

heck a xenobiologist can probably make sentient carp to kill the portal, non space dragon carp are indestinguishable from the dragon`s carp except for their higher speed AND they have way more obj damage than the dragon`s carp.

File: code/modules/vehicles/mecha/mech_melee_attack.dm [hunk 4]
```diff
@@ -42,7 +42,14 @@
 			return 0
 	mecha_attacker.visible_message(span_danger("[mecha_attacker] hits [src]!"), span_danger("You hit [src]!"), null, COMBAT_MESSAGE_RANGE)
 	..()
-	return take_damage(mecha_attacker.force * 3, mecha_attacker.damtype, "melee", FALSE, get_dir(src, mecha_attacker)) // multiplied by 3 so we can hit objs hard but not be overpowered against mobs.
+	return take_damage(mecha_attacker.force * get_obj_mecha_damage_multiplier(), mecha_attacker.damtype, "melee", FALSE, get_dir(src, mecha_attacker))
+
+/** Gets the object mecha damage multiplier.
+ * Some objects (carp rift) shouldn't have a high multiplier, override this proc if necessary.
+ */
+/obj/proc/get_obj_mecha_damage_multiplier(obj/vehicle/sealed/mecha/mecha_attacker, mob/living/user)
```
Comment (optimumtact) [inline code/modules/vehicles/mecha/mech_melee_attack.dm:50]: you added a proc to literally every object in the game for a multiplier on a single object.

no, use a signal with a return value or do not pass go.

**Comment by optimumtact** [comment]
this doesn't seem like a balance change worth the effort tbh.

File: code/modules/vehicles/mecha/mech_melee_attack.dm (hunk 4 above)
Comment (MrMelbert) [reply code/modules/vehicles/mecha/mech_melee_attack.dm:50]: Signals can't return anything besides a bitflag

Comment (MrDoomBringer) [reply code/modules/vehicles/mecha/mech_melee_attack.dm:50]: Can have it return a bitflag with a value of 3 lol

Comment (optimumtact) [reply code/modules/vehicles/mecha/mech_melee_attack.dm:50]: > Signals can't return anything besides a bitflag

bro you are tripped

Comment (Fikou) [reply code/modules/vehicles/mecha/mech_melee_attack.dm:50]: > Can have it return a bitflag with a value of 3 lol

you can physically return a bitflag but you are not intended to it goes against the design of signals. they are not meant to be used as proc calls

Comment (MrMelbert) [reply code/modules/vehicles/mecha/mech_melee_attack.dm:50]: You could pass the signal an empty list and allow consumers to add modifiers to the list

```dm
var/list/damage_mods = list(3)
SEND_SIGNAL(src, COMSIG_OBJ_MECHA_HIT, mecha_attacker, user, damage_mods)

var/final_damage = mecha_attacker.force
for(var/newmod in damage_mods)
    final_damage *= newmod
```

and

```
/obj/carp_rift/proc/on_mecha_hit(obj/item/source, obj/vechicle/mecha/attacker, mob/living/attacker_mob, list/damage_mods)
    SIGNAL_HANDLER
    damage_mods += 0.25 // 3 (base amount) * 0.25 = 0.27

```

Just a concept

**Pull request closed without merging**